_docs/implementation.md

```diff
@@ -0,0 +1,121 @@
+# Introduction
+This proposal outlines a method for injecting runtime hooks into target functions in Go programs, enabling dynamic monitoring and modification of function behavior. The approach leverages trampoline code injection and function pointer redirection to seamlessly integrate monitoring logic without requiring significant changes to the user's codebase.
+
+The goal is to provide a flexible and non-intrusive mechanism for instrumenting Go applications, particularly for use cases such as observability (e.g., OpenTelemetry), debugging, or performance profiling.
+
+# Core Principles
+## 1. Trampoline Code Injection
+Trampoline function calls are injected into the Target (lib-side) function, which ultimately jumps to the actual hook code via the function pointer Hook.
+
+```go
+func Target() {
+    Trampoline()
+    ....
+}
+
+func Trampoline() {
+    Hook()
+}
+
+//go:linkname Hook github.com/open-telemetry/opentelemetry-go-compile-instrumentation/sdk/hook.MyHook
+var Hook func()
+```
+The two-level indirection (trampoline call to hook) is a deliberate design choice that offers several key benefits:
+
+- Exception Handling: The trampoline catches panics and isolates exception handling, preventing them from affecting the target function or hook code.
+- Context Construction: The trampoline initializes and manages the necessary context before invoking the hook code.
+- Decoupling: The trampoline decouples the hook code from the target function, enabling flexibility and maintainability.
+- Debugging: The trampoline provides a centralized point for debugging and observability.
+
+
+## 2. Linkage via golinkname
+The `Hook` function is linked to the monitoring code using the `//go:linkname` directive. This allows us to dynamically associate the hook with the target function at compile time.
```

Comment:
I'd clarify this is putting us in the "handshake" style of `//go:linkname` (this is using language from the maintainers), which is expected to continue being supported indefinitely (as opposed to "pull", which is blocked outside of grandfathered-in exceptions)